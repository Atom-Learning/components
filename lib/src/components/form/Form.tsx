import React from 'react'
import {
  DefaultValues,
  FieldErrors,
  FieldValues,
  FormProvider,
  Mode,
  useForm,
  UseFormMethods
} from 'react-hook-form'

import { styled } from '~/stitches'

const StyledForm = styled('form', {})

type StyledFormProps = Omit<
  React.ComponentPropsWithoutRef<typeof StyledForm>,
  'onSubmit' | 'onError'
>

interface FormProps<TFormData extends FieldValues> extends StyledFormProps {
  defaultValues?: DefaultValues<TFormData>
  validationMode?: Mode
  onSubmit: (data: TFormData) => void | any
  onError?: (errors: FieldErrors<TFormData>) => void
  children:
    | React.ReactNode
    | ((methods: UseFormMethods<TFormData>) => React.ReactNode)
}

export const Form = <TFormData extends FieldValues>(
  props: FormProps<TFormData>
) => {
  const {
    children,
    defaultValues,
    validationMode = 'onBlur',
    onSubmit,
    onError,
    ...rest
  } = props

  const methods = useForm<TFormData>({
    defaultValues,
    mode: validationMode
  })

  return (
    <FormProvider {...methods}>
      <StyledForm
        aria-label="form"
        onSubmit={methods.handleSubmit(onSubmit, onError)}
        {...rest}
      >
        {typeof children === 'function' ? children(methods) : children}
      </StyledForm>
    </FormProvider>
  )
}

Form.displayName = 'Form'
