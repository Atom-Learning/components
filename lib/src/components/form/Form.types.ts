import type { Mode, UseFormMethods } from 'react-hook-form'

export enum StorageEnum {
  LOCAL = 'local',
  SESSION = 'session'
}

type ExcludeIncludeConfig = {
  exclude?: string[]
  include?: string[]
}

export type PersistOptions = {
  id: string
  storage?: StorageEnum
} & ExcludeIncludeConfig

export type FormPersistParams = {
  storage: Storage
} & ExcludeIncludeConfig

export type PersistFormWrapperValues = {
  persist: PersistOptions
  watch
  setValue
  children: React.ReactNode | any
}

export type FormContentValues = {
  formMethods
  handleSubmit: (
    submitHandler: (data: any) => void | any,
    submitErrorHandler?: (errors: any) => void
  ) => (e: any) => Promise<void>
  onSubmit: (data: any) => void | any
  onError?: (errors: any) => void
  children: React.ReactNode | any
}

export type FormValues = {
  defaultValues?: { [key: string]: string | number }
  onSubmit: (data: any) => void | any
  onError?: (errors: any) => void
  criteriaMode?: 'firstError' | 'all'
  validationMode?: Mode
  persist?: PersistOptions
} & (
  | { children: React.ReactNode; render?: never }
  | { children?: never; render: (methods: UseFormMethods) => React.ReactNode }
)
