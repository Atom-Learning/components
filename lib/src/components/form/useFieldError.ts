import * as React from 'react'
import delve from 'dlv'
import { useFormContext } from 'react-hook-form'

export const useFieldError = (
  fieldName: string,
  validationMode: 'firstError' | 'all' = 'firstError'
) => {
  const { errors } = useFormContext()

  const fieldErrors = getFieldErrors(errors, fieldName, validationMode)

  return fieldErrors
}

const getFieldErrors = (
  errors,
  fieldName,
  validationMode
): Array<string> | null => {
  // in case of migration to a newer version of react-hook-form the path needs to updated to account for dynamic fields naming convention
  const fieldPath = fieldName.split(/[,[\].]+?/).filter(Boolean)

  // const getErrors = () => delve(errors, fieldPath)
  const error = delve(errors, fieldPath)

  // `message` is the first validation error, as prioritised by react-hook-form.
  // `types` is an object containing all current validation errors, keyed by the validation type
  const { message, types } = error || {}

  if (!types && !message) return null

  if (validationMode === 'firstError' && message) return [message]

  return Object.values(types)
}
