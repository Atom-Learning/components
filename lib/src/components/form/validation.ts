export type ValidationOptions = {
  required?: Option<boolean> | string
  min?: Option<number>
  max?: Option<number>
  criteriaMode?: 'all' | 'firstError'
  minLength?: Option<number>
  maxLength?: Option<number>
  pattern?: Option<RegExp>
  validate?: Validator | { [key: string]: Validator }
  valueAsNumber?: boolean
  valueAsDate?: boolean
  process?: Processor
}

export type ValidationError = {
  message: string
  type:
    | 'required'
    | 'min'
    | 'max'
    | 'minLength'
    | 'maxLength'
    | 'pattern'
    | 'validate'
}

// Returns `true` if a value passes validation, or a string with the error if not
type Validator = (value: any) => boolean | string

// Processes a value after validation
type Processor = (value: any) => any

type Option<T> = {
  value: T
  message: string
}
