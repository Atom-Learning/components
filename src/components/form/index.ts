export { Form } from './Form'
export { ValidationOptions } from './validation'
