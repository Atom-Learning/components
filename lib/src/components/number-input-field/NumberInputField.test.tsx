import { render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { axe } from 'jest-axe'
import * as React from 'react'

import { Form } from '../form'
import { NumberInputField } from '.'
import type { NumberInputFieldProps } from './NumberInputField'

const renderComponent = (props: Partial<NumberInputFieldProps> = {}) => {
  return render(
    <Form onSubmit={() => console.log('Form submitted')}>
      <NumberInputField
        name="numberInputField"
        label="Number Input"
        {...props}
      />
      <button type="submit">Submit</button>
    </Form>
  )
}

describe(`NumberInputField component`, () => {
  it('renders a field with a number input - has no programmatically detectable a11y issues', async () => {
    const { container } = renderComponent()

    expect(container).toMatchSnapshot()
    expect(await axe(container)).toHaveNoViolations()
  })

  it('renders a field with a disabled input - has no programmatically detectable a11y issues', async () => {
    const { container, getByRole } = renderComponent({ isDisabled: true })

    const input = getByRole('spinbutton')

    expect(container).toMatchSnapshot()
    expect(input).toHaveAttribute('disabled')
    expect(await axe(container)).toHaveNoViolations()
  })

  it('renders a field in error state - has no programmatically detectable a11y issues', async () => {
    const errorText = 'Cannot enter values less than 7'

    const { container, findByText, getByRole } = renderComponent({
      validation: { min: { value: 7, message: errorText } }
    })

    userEvent.click(getByRole('button'))
    await findByText(errorText)

    expect(container).toMatchSnapshot()
    expect(await axe(container)).toHaveNoViolations()
  })
})
