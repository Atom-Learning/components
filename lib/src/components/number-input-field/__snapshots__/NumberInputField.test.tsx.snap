// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberInputField component renders a field in error state - has no programmatically detectable a11y issues 1`] = `
@media  {
  .c-dhzjXW {
    display: flex;
  }

  .c-cVSpVL {
    color: var(--colors-tonal500);
    font-family: var(--fonts-body);
    margin: 0;
  }

  .c-jVLVTy {
    z-index: 1;
  }

  .c-oUmPa {
    align-items: center;
    -webkit-appearance: none;
    appearance: none;
    background: white;
    border: unset;
    border-radius: var(--radii-0);
    box-sizing: border-box;
    cursor: pointer;
    display: flex;
    flex-shrink: 0;
    justify-content: center;
    padding: unset;
    transition: all 100ms ease-out;
  }

  .c-oUmPa[disabled] {
    opacity: 0.3;
    cursor: not-allowed;
  }

  .c-hMRxhp {
    display: inline-block;
    fill: none;
    flex-shrink: 0;
    stroke: currentcolor;
    stroke-linecap: round;
    stroke-linejoin: round;
    vertical-align: middle;
  }

  .c-dWvLOH {
    border-color: var(--colors-tonal400) !important;
    z-index: 1;
    height: 100% !important;
  }

  .c-dWvLOH:hover {
    background: var(--colors-tonal50);
  }

  .c-dWvLOH:hover svg {
    color: var(--colors-tonal400);
  }

  .c-dWvLOH svg {
    color: var(--colors-tonal300);
  }

  .c-dWvLOH:active {
    background: var(--colors-tonal100);
  }

  .c-dWvLOH:active svg {
    color: var(--colors-tonal500);
  }

  .c-dWvLOH:disabled {
    opacity: 0.3;
    pointer-events: none;
  }

  .c-hVoQtf {
    background-color: var(--colors-tonal500);
    border-radius: var(--radii-0);
    box-shadow: var(--shadows-0);
    color: white;
    font-family: var(--fonts-body);
    font-size: var(--fontSizes-sm);
    line-height: 1.5;
    white-space: normal;
    padding-left: var(--space-3);
    padding-right: var(--space-3);
    padding-top: var(--space-2);
    padding-bottom: var(--space-2);
    z-index: 10;
  }

@media (prefers-reduced-motion: no-preference) {
    .c-hVoQtf {
      animation-duration: 75ms;
      animation-timing-function: cubic-bezier(0.16, 1, 0.3, 1);
      will-change: transform, opacity;
    }
}

@media (prefers-reduced-motion: no-preference) {
    .c-hVoQtf[data-state="delayed-open"][data-side="top"] {
      animation-name: k-ftcNPK;
    }
}

@media (prefers-reduced-motion: no-preference) {
    .c-hVoQtf[data-state="delayed-open"][data-side="right"] {
      animation-name: k-hbaHED;
    }
}

@media (prefers-reduced-motion: no-preference) {
    .c-hVoQtf[data-state="delayed-open"][data-side="bottom"] {
      animation-name: k-daNevH;
    }
}

@media (prefers-reduced-motion: no-preference) {
    .c-hVoQtf[data-state="delayed-open"][data-side="left"] {
      animation-name: k-fvyGIa;
    }
}

  .c-lfQsGQ {
    -webkit-appearance: none;
    appearance: none;
    border: 1px solid var(--colors-tonal400);
    border-radius: var(--radii-0);
    box-shadow: none;
    box-sizing: border-box;
    color: var(--colors-tonal600);
    cursor: text;
    display: block;
    font-family: var(--fonts-body);
    padding-left: var(--space-3);
    padding-right: var(--space-3);
    transition: all 100ms ease-out;
    width: 100%;
  }

  .c-lfQsGQ:focus {
    border-color: var(--colors-primary800);
    outline: none;
  }

  .c-lfQsGQ[disabled] {
    opacity: 0.3;
    cursor: not-allowed;
  }

  .c-lfQsGQ::placeholder {
    color: var(--colors-tonal300);
    opacity: 1;
  }

  .c-dyvMgW {
    font-family: var(--fonts-body);
    font-weight: 400;
    margin: 0;
  }

  .c-dyvMgW > .c-dyvMgW:before,
  .c-dyvMgW > .c-dyvMgW:after {
    display: none;
  }
}

@media  {
  .c-cVSpVL-gvmVBy-size-md {
    font-size: var(--fontSizes-md);
    line-height: 1.5;
  }

  .c-cVSpVL-gvmVBy-size-md::before {
    content: '';
    margin-bottom: -0.3864em;
    display: table;
  }

  .c-cVSpVL-gvmVBy-size-md::after {
    content: '';
    margin-top: -0.3864em;
    display: table;
  }

  .c-cVSpVL-grKRUr-type-block {
    display: block;
    font-weight: 600;
  }

  .c-oUmPa-PrXKS-size-md {
    height: var(--sizes-4);
    width: var(--sizes-4);
  }

  .c-hMRxhp-dMrjnF-size-md {
    height: var(--sizes-2);
    width: var(--sizes-2);
    stroke-width: 1.75;
  }

  .c-hVoQtf-gWQnqe-size-md {
    max-width: 250px;
  }

  .c-lfQsGQ-fwoCmE-size-md {
    height: var(--sizes-4);
    font-size: var(--fontSizes-md);
    line-height: 2;
  }

  .c-lfQsGQ-bXVTjQ-state-error {
    border: 1px solid var(--colors-danger);
  }

  .c-PJLV-iymbh-theme-error {
    color: var(--colors-danger);
  }

  .c-hMRxhp-cyeZeh-size-sm {
    height: var(--sizes-1);
    width: var(--sizes-1);
    stroke-width: 1.5;
  }

  .c-dyvMgW-bndJoy-size-sm {
    font-size: var(--fontSizes-sm);
    line-height: 1.53;
  }

  .c-dyvMgW-bndJoy-size-sm::before {
    content: '';
    margin-bottom: -0.4056em;
    display: table;
  }

  .c-dyvMgW-bndJoy-size-sm::after {
    content: '';
    margin-top: -0.4056em;
    display: table;
  }
}

@media  {
  .c-oUmPa-fboUUB-cv {
    border: 1px solid;
    border-color: currentColor;
    color: var(--colors-primary800);
  }

  .c-oUmPa-fboUUB-cv:not(:disabled):hover,
  .c-oUmPa-fboUUB-cv:not(:disabled):focus {
    color: var(--colors-primary900);
  }

  .c-oUmPa-fboUUB-cv:not(:disabled):active {
    color: var(--colors-primary1000);
  }
}

@media  {
  .c-dhzjXW-iwHglP-css {
    justify-content: space-between;
    align-items: center;
    margin-bottom: var(--space-3);
  }

  .c-dWvLOH-iwWtoZ-css {
    border-right: none;
    border-top-right-radius: 0px;
    border-bottom-right-radius: 0px;
  }

  .c-lfQsGQ-ibwGleD-css {
    border-radius: 0px;
    width: var(--sizes-6);
  }

  .c-lfQsGQ-ibwGleD-css:disabled {
    opacity: 0.3;
    pointer-events: none;
  }

  .c-dWvLOH-ihgsOSq-css {
    border-left: none;
    border-top-left-radius: 0px;
    border-bottom-left-radius: 0px;
  }

  .c-dhzjXW-ijkHUQU-css {
    margin-top: var(--space-2);
  }

  .c-hMRxhp-ijkHsGW-css {
    margin-right: var(--space-2);
  }

  .c-dyvMgW-iiTbKsx-css {
    transform: translateY(var(--space-0));
  }
}

<div>
  <form
    aria-label="form"
    class="c-PJLV"
  >
    <div
      class="c-PJLV"
    >
      <div
        class="c-dhzjXW c-dhzjXW-iwHglP-css"
      >
        <label
          class="c-cVSpVL c-cVSpVL-gvmVBy-size-md c-cVSpVL-grKRUr-type-block"
          for="numberInputField"
        >
          Number Input
        </label>
      </div>
      <div
        class="c-dhzjXW"
      >
        <span
          class="c-jVLVTy c-PJLV"
          data-state="closed"
          tabindex="-1"
        >
          <button
            aria-label="decrement"
            class="c-oUmPa c-oUmPa-PrXKS-size-md c-oUmPa-fboUUB-cv c-dWvLOH c-dWvLOH-iwWtoZ-css"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="c-hMRxhp c-hMRxhp-dMrjnF-size-md c-hMRxhp-iPJLV-css"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M20 12H4"
              />
            </svg>
          </button>
        </span>
        <input
          aria-invalid="true"
          aria-valuemax="9007199254740991"
          aria-valuemin="0"
          aria-valuenow="0"
          class="c-lfQsGQ c-lfQsGQ-fwoCmE-size-md c-lfQsGQ-bXVTjQ-state-error c-lfQsGQ-ibwGleD-css"
          id="numberInputField"
          inputmode="numeric"
          name="numberInputField"
          pattern="[0-9]*"
          role="spinbutton"
          type="text"
          value="0"
        />
        <span
          class="c-jVLVTy c-PJLV"
          data-state="closed"
          tabindex="-1"
        >
          <button
            aria-label="increment"
            class="c-oUmPa c-oUmPa-PrXKS-size-md c-oUmPa-fboUUB-cv c-dWvLOH c-dWvLOH-ihgsOSq-css"
            tabindex="-1"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="c-hMRxhp c-hMRxhp-dMrjnF-size-md c-hMRxhp-iPJLV-css"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M20 12H4m8-8v16"
              />
            </svg>
          </button>
        </span>
      </div>
      <div
        class="c-dhzjXW c-PJLV c-PJLV-iymbh-theme-error c-dhzjXW-ijkHUQU-css"
      >
        <svg
          aria-hidden="true"
          class="c-hMRxhp c-hMRxhp-cyeZeh-size-sm c-hMRxhp-ijkHsGW-css"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 8v5"
          />
          <line
            x1="12"
            x2="12"
            y1="16"
            y2="16"
          />
          <circle
            cx="12"
            cy="12"
            r="10"
          />
        </svg>
        <span
          class="c-dyvMgW c-dyvMgW-bndJoy-size-sm c-dyvMgW-iiTbKsx-css"
        >
          Cannot enter values less than 7
        </span>
      </div>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`NumberInputField component renders a field with a disabled input - has no programmatically detectable a11y issues 1`] = `
@media  {
  .c-dhzjXW {
    display: flex;
  }

  .c-cVSpVL {
    color: var(--colors-tonal500);
    font-family: var(--fonts-body);
    margin: 0;
  }

  .c-jVLVTy {
    z-index: 1;
  }

  .c-oUmPa {
    align-items: center;
    -webkit-appearance: none;
    appearance: none;
    background: white;
    border: unset;
    border-radius: var(--radii-0);
    box-sizing: border-box;
    cursor: pointer;
    display: flex;
    flex-shrink: 0;
    justify-content: center;
    padding: unset;
    transition: all 100ms ease-out;
  }

  .c-oUmPa[disabled] {
    opacity: 0.3;
    cursor: not-allowed;
  }

  .c-hMRxhp {
    display: inline-block;
    fill: none;
    flex-shrink: 0;
    stroke: currentcolor;
    stroke-linecap: round;
    stroke-linejoin: round;
    vertical-align: middle;
  }

  .c-dWvLOH {
    border-color: var(--colors-tonal400) !important;
    z-index: 1;
    height: 100% !important;
  }

  .c-dWvLOH:hover {
    background: var(--colors-tonal50);
  }

  .c-dWvLOH:hover svg {
    color: var(--colors-tonal400);
  }

  .c-dWvLOH svg {
    color: var(--colors-tonal300);
  }

  .c-dWvLOH:active {
    background: var(--colors-tonal100);
  }

  .c-dWvLOH:active svg {
    color: var(--colors-tonal500);
  }

  .c-dWvLOH:disabled {
    opacity: 0.3;
    pointer-events: none;
  }

  .c-hVoQtf {
    background-color: var(--colors-tonal500);
    border-radius: var(--radii-0);
    box-shadow: var(--shadows-0);
    color: white;
    font-family: var(--fonts-body);
    font-size: var(--fontSizes-sm);
    line-height: 1.5;
    white-space: normal;
    padding-left: var(--space-3);
    padding-right: var(--space-3);
    padding-top: var(--space-2);
    padding-bottom: var(--space-2);
    z-index: 10;
  }

@media (prefers-reduced-motion: no-preference) {
    .c-hVoQtf {
      animation-duration: 75ms;
      animation-timing-function: cubic-bezier(0.16, 1, 0.3, 1);
      will-change: transform, opacity;
    }
}

@media (prefers-reduced-motion: no-preference) {
    .c-hVoQtf[data-state="delayed-open"][data-side="top"] {
      animation-name: k-ftcNPK;
    }
}

@media (prefers-reduced-motion: no-preference) {
    .c-hVoQtf[data-state="delayed-open"][data-side="right"] {
      animation-name: k-hbaHED;
    }
}

@media (prefers-reduced-motion: no-preference) {
    .c-hVoQtf[data-state="delayed-open"][data-side="bottom"] {
      animation-name: k-daNevH;
    }
}

@media (prefers-reduced-motion: no-preference) {
    .c-hVoQtf[data-state="delayed-open"][data-side="left"] {
      animation-name: k-fvyGIa;
    }
}

  .c-lfQsGQ {
    -webkit-appearance: none;
    appearance: none;
    border: 1px solid var(--colors-tonal400);
    border-radius: var(--radii-0);
    box-shadow: none;
    box-sizing: border-box;
    color: var(--colors-tonal600);
    cursor: text;
    display: block;
    font-family: var(--fonts-body);
    padding-left: var(--space-3);
    padding-right: var(--space-3);
    transition: all 100ms ease-out;
    width: 100%;
  }

  .c-lfQsGQ:focus {
    border-color: var(--colors-primary800);
    outline: none;
  }

  .c-lfQsGQ[disabled] {
    opacity: 0.3;
    cursor: not-allowed;
  }

  .c-lfQsGQ::placeholder {
    color: var(--colors-tonal300);
    opacity: 1;
  }
}

@media  {
  .c-cVSpVL-gvmVBy-size-md {
    font-size: var(--fontSizes-md);
    line-height: 1.5;
  }

  .c-cVSpVL-gvmVBy-size-md::before {
    content: '';
    margin-bottom: -0.3864em;
    display: table;
  }

  .c-cVSpVL-gvmVBy-size-md::after {
    content: '';
    margin-top: -0.3864em;
    display: table;
  }

  .c-cVSpVL-grKRUr-type-block {
    display: block;
    font-weight: 600;
  }

  .c-oUmPa-PrXKS-size-md {
    height: var(--sizes-4);
    width: var(--sizes-4);
  }

  .c-hMRxhp-dMrjnF-size-md {
    height: var(--sizes-2);
    width: var(--sizes-2);
    stroke-width: 1.75;
  }

  .c-hVoQtf-gWQnqe-size-md {
    max-width: 250px;
  }

  .c-lfQsGQ-fwoCmE-size-md {
    height: var(--sizes-4);
    font-size: var(--fontSizes-md);
    line-height: 2;
  }
}

@media  {
  .c-oUmPa-fboUUB-cv {
    border: 1px solid;
    border-color: currentColor;
    color: var(--colors-primary800);
  }

  .c-oUmPa-fboUUB-cv:not(:disabled):hover,
  .c-oUmPa-fboUUB-cv:not(:disabled):focus {
    color: var(--colors-primary900);
  }

  .c-oUmPa-fboUUB-cv:not(:disabled):active {
    color: var(--colors-primary1000);
  }
}

@media  {
  .c-dhzjXW-iwHglP-css {
    justify-content: space-between;
    align-items: center;
    margin-bottom: var(--space-3);
  }

  .c-dWvLOH-iwWtoZ-css {
    border-right: none;
    border-top-right-radius: 0px;
    border-bottom-right-radius: 0px;
  }

  .c-lfQsGQ-ibwGleD-css {
    border-radius: 0px;
    width: var(--sizes-6);
  }

  .c-lfQsGQ-ibwGleD-css:disabled {
    opacity: 0.3;
    pointer-events: none;
  }

  .c-dWvLOH-ihgsOSq-css {
    border-left: none;
    border-top-left-radius: 0px;
    border-bottom-left-radius: 0px;
  }
}

<div>
  <form
    aria-label="form"
    class="c-PJLV"
  >
    <div
      class="c-PJLV"
    >
      <div
        class="c-dhzjXW c-dhzjXW-iwHglP-css"
      >
        <label
          class="c-cVSpVL c-cVSpVL-gvmVBy-size-md c-cVSpVL-grKRUr-type-block"
          for="numberInputField"
        >
          Number Input
        </label>
      </div>
      <div
        class="c-dhzjXW"
      >
        <span
          class="c-jVLVTy c-PJLV"
          data-state="closed"
          tabindex="-1"
        >
          <button
            aria-label="decrement"
            class="c-oUmPa c-oUmPa-PrXKS-size-md c-oUmPa-fboUUB-cv c-dWvLOH c-dWvLOH-iwWtoZ-css"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="c-hMRxhp c-hMRxhp-dMrjnF-size-md c-hMRxhp-iPJLV-css"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M20 12H4"
              />
            </svg>
          </button>
        </span>
        <input
          aria-valuemax="9007199254740991"
          aria-valuemin="0"
          aria-valuenow="0"
          class="c-lfQsGQ c-lfQsGQ-fwoCmE-size-md c-lfQsGQ-ibwGleD-css"
          disabled=""
          id="numberInputField"
          inputmode="numeric"
          name="numberInputField"
          pattern="[0-9]*"
          role="spinbutton"
          type="text"
          value="0"
        />
        <span
          class="c-jVLVTy c-PJLV"
          data-state="closed"
          tabindex="-1"
        >
          <button
            aria-label="increment"
            class="c-oUmPa c-oUmPa-PrXKS-size-md c-oUmPa-fboUUB-cv c-dWvLOH c-dWvLOH-ihgsOSq-css"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="c-hMRxhp c-hMRxhp-dMrjnF-size-md c-hMRxhp-iPJLV-css"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M20 12H4m8-8v16"
              />
            </svg>
          </button>
        </span>
      </div>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;

exports[`NumberInputField component renders a field with a number input - has no programmatically detectable a11y issues 1`] = `
@media  {
  .c-dhzjXW {
    display: flex;
  }

  .c-cVSpVL {
    color: var(--colors-tonal500);
    font-family: var(--fonts-body);
    margin: 0;
  }

  .c-jVLVTy {
    z-index: 1;
  }

  .c-oUmPa {
    align-items: center;
    -webkit-appearance: none;
    appearance: none;
    background: white;
    border: unset;
    border-radius: var(--radii-0);
    box-sizing: border-box;
    cursor: pointer;
    display: flex;
    flex-shrink: 0;
    justify-content: center;
    padding: unset;
    transition: all 100ms ease-out;
  }

  .c-oUmPa[disabled] {
    opacity: 0.3;
    cursor: not-allowed;
  }

  .c-hMRxhp {
    display: inline-block;
    fill: none;
    flex-shrink: 0;
    stroke: currentcolor;
    stroke-linecap: round;
    stroke-linejoin: round;
    vertical-align: middle;
  }

  .c-dWvLOH {
    border-color: var(--colors-tonal400) !important;
    z-index: 1;
    height: 100% !important;
  }

  .c-dWvLOH:hover {
    background: var(--colors-tonal50);
  }

  .c-dWvLOH:hover svg {
    color: var(--colors-tonal400);
  }

  .c-dWvLOH svg {
    color: var(--colors-tonal300);
  }

  .c-dWvLOH:active {
    background: var(--colors-tonal100);
  }

  .c-dWvLOH:active svg {
    color: var(--colors-tonal500);
  }

  .c-dWvLOH:disabled {
    opacity: 0.3;
    pointer-events: none;
  }

  .c-hVoQtf {
    background-color: var(--colors-tonal500);
    border-radius: var(--radii-0);
    box-shadow: var(--shadows-0);
    color: white;
    font-family: var(--fonts-body);
    font-size: var(--fontSizes-sm);
    line-height: 1.5;
    white-space: normal;
    padding-left: var(--space-3);
    padding-right: var(--space-3);
    padding-top: var(--space-2);
    padding-bottom: var(--space-2);
    z-index: 10;
  }

@media (prefers-reduced-motion: no-preference) {
    .c-hVoQtf {
      animation-duration: 75ms;
      animation-timing-function: cubic-bezier(0.16, 1, 0.3, 1);
      will-change: transform, opacity;
    }
}

@media (prefers-reduced-motion: no-preference) {
    .c-hVoQtf[data-state="delayed-open"][data-side="top"] {
      animation-name: k-ftcNPK;
    }
}

@media (prefers-reduced-motion: no-preference) {
    .c-hVoQtf[data-state="delayed-open"][data-side="right"] {
      animation-name: k-hbaHED;
    }
}

@media (prefers-reduced-motion: no-preference) {
    .c-hVoQtf[data-state="delayed-open"][data-side="bottom"] {
      animation-name: k-daNevH;
    }
}

@media (prefers-reduced-motion: no-preference) {
    .c-hVoQtf[data-state="delayed-open"][data-side="left"] {
      animation-name: k-fvyGIa;
    }
}

  .c-lfQsGQ {
    -webkit-appearance: none;
    appearance: none;
    border: 1px solid var(--colors-tonal400);
    border-radius: var(--radii-0);
    box-shadow: none;
    box-sizing: border-box;
    color: var(--colors-tonal600);
    cursor: text;
    display: block;
    font-family: var(--fonts-body);
    padding-left: var(--space-3);
    padding-right: var(--space-3);
    transition: all 100ms ease-out;
    width: 100%;
  }

  .c-lfQsGQ:focus {
    border-color: var(--colors-primary800);
    outline: none;
  }

  .c-lfQsGQ[disabled] {
    opacity: 0.3;
    cursor: not-allowed;
  }

  .c-lfQsGQ::placeholder {
    color: var(--colors-tonal300);
    opacity: 1;
  }
}

@media  {
  .c-cVSpVL-gvmVBy-size-md {
    font-size: var(--fontSizes-md);
    line-height: 1.5;
  }

  .c-cVSpVL-gvmVBy-size-md::before {
    content: '';
    margin-bottom: -0.3864em;
    display: table;
  }

  .c-cVSpVL-gvmVBy-size-md::after {
    content: '';
    margin-top: -0.3864em;
    display: table;
  }

  .c-cVSpVL-grKRUr-type-block {
    display: block;
    font-weight: 600;
  }

  .c-oUmPa-PrXKS-size-md {
    height: var(--sizes-4);
    width: var(--sizes-4);
  }

  .c-hMRxhp-dMrjnF-size-md {
    height: var(--sizes-2);
    width: var(--sizes-2);
    stroke-width: 1.75;
  }

  .c-hVoQtf-gWQnqe-size-md {
    max-width: 250px;
  }

  .c-lfQsGQ-fwoCmE-size-md {
    height: var(--sizes-4);
    font-size: var(--fontSizes-md);
    line-height: 2;
  }
}

@media  {
  .c-oUmPa-fboUUB-cv {
    border: 1px solid;
    border-color: currentColor;
    color: var(--colors-primary800);
  }

  .c-oUmPa-fboUUB-cv:not(:disabled):hover,
  .c-oUmPa-fboUUB-cv:not(:disabled):focus {
    color: var(--colors-primary900);
  }

  .c-oUmPa-fboUUB-cv:not(:disabled):active {
    color: var(--colors-primary1000);
  }
}

@media  {
  .c-dhzjXW-iwHglP-css {
    justify-content: space-between;
    align-items: center;
    margin-bottom: var(--space-3);
  }

  .c-dWvLOH-iwWtoZ-css {
    border-right: none;
    border-top-right-radius: 0px;
    border-bottom-right-radius: 0px;
  }

  .c-lfQsGQ-ibwGleD-css {
    border-radius: 0px;
    width: var(--sizes-6);
  }

  .c-lfQsGQ-ibwGleD-css:disabled {
    opacity: 0.3;
    pointer-events: none;
  }

  .c-dWvLOH-ihgsOSq-css {
    border-left: none;
    border-top-left-radius: 0px;
    border-bottom-left-radius: 0px;
  }
}

<div>
  <form
    aria-label="form"
    class="c-PJLV"
  >
    <div
      class="c-PJLV"
    >
      <div
        class="c-dhzjXW c-dhzjXW-iwHglP-css"
      >
        <label
          class="c-cVSpVL c-cVSpVL-gvmVBy-size-md c-cVSpVL-grKRUr-type-block"
          for="numberInputField"
        >
          Number Input
        </label>
      </div>
      <div
        class="c-dhzjXW"
      >
        <span
          class="c-jVLVTy c-PJLV"
          data-state="closed"
          tabindex="-1"
        >
          <button
            aria-label="decrement"
            class="c-oUmPa c-oUmPa-PrXKS-size-md c-oUmPa-fboUUB-cv c-dWvLOH c-dWvLOH-iwWtoZ-css"
            disabled=""
            tabindex="-1"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="c-hMRxhp c-hMRxhp-dMrjnF-size-md c-hMRxhp-iPJLV-css"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M20 12H4"
              />
            </svg>
          </button>
        </span>
        <input
          aria-valuemax="9007199254740991"
          aria-valuemin="0"
          aria-valuenow="0"
          class="c-lfQsGQ c-lfQsGQ-fwoCmE-size-md c-lfQsGQ-ibwGleD-css"
          id="numberInputField"
          inputmode="numeric"
          name="numberInputField"
          pattern="[0-9]*"
          role="spinbutton"
          type="text"
          value="0"
        />
        <span
          class="c-jVLVTy c-PJLV"
          data-state="closed"
          tabindex="-1"
        >
          <button
            aria-label="increment"
            class="c-oUmPa c-oUmPa-PrXKS-size-md c-oUmPa-fboUUB-cv c-dWvLOH c-dWvLOH-ihgsOSq-css"
            tabindex="-1"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="c-hMRxhp c-hMRxhp-dMrjnF-size-md c-hMRxhp-iPJLV-css"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M20 12H4m8-8v16"
              />
            </svg>
          </button>
        </span>
      </div>
    </div>
    <button
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
